validator/sawtooth_validator/consensus/registry.py

```diff
@@ -45,27 +46,38 @@ def has_active_engine(self):
         with self._lock:
             return self._active is not None
 
-    def register_engine(self, connection_id, name, version):
+    def register_engine(self, connection_id, name, version,
+                        additional_protocols):
         with self._lock:
-            # If this engine is already registered, remove the old connection
+            # If this engine is already registered, remove the old connection.
+            # If there is an existing engine that supports one of the same
+            # protocols as this engine, unregister it.
```

Comment:
The biggest open question for me is the mechanics of this; what should we do if another engine gets unregistered? Should we notify it? What if it was the active engine, and this engine that displaced it doesn't support the active protocol?

Ultimately, this is up to the sysadmin; really you shouldn't be using overlapping protocols unless something strictly replaces another engine (e.g. an upgrade). Maybe we could enforce that in order to replace an engine, the new one must support a superset of the displaced engine's protocols?

Reply:
Updated so that more than one engine can support a given protocol; the engine that registered first will become active if that protocol is activated. If a registering engine supports all of the same protocols as an existing registered engine (protocols are a superset), then the existing engine is unregistered (this is deemed to be an upgrade).